xc/mem_stats.ml

```diff
@@ -0,0 +1,331 @@
+
+  (** [parse_value s] parses values of the form '  N kB' or '    N'. *)
+  let parse_value s =
+    s |> Astring.String.trim |> drop_unit |> Int64.of_string
+
+  let file_lines_filter_map f ~path =
+    let fold acc line =
+      match f line with
+      | None -> acc
+      | Some (k, v) -> KV.add k v acc
+    in
+    file_lines_fold fold KV.empty path
+
+  open SimpleMetrics
+  let parse_keyvalue fields ~path () =
+    let parse_pairs = function
+      | None -> None
+      | Some (key, value) ->
+          match KV.find key fields with
+          | None -> None
+          | Some ds ->
+              let v = parse_value value in
+              Some (key, ds_update ds @@ ds_int64 v)
+    in
+    file_lines_filter_map ~path @@ fun line ->
+    line |> Astring.String.cut ~sep:":" |> parse_pairs
+
+  let define_fields ~path l =
+    let kv = ListLabels.fold_left l ~init:KV.empty ~f:(fun acc (key, ds) ->
+      KV.add key ds acc) in
+    parse_keyvalue kv ~path:(Filename.concat "/proc/self" path)
+
+  let unknown = ds_int64 (-1L)
+  let kib ?default key desc = key, kib ?default key desc unknown
+  let count ?default key desc = key, count ?default key desc unknown
+
+  module Fields = struct
+    let vmdata = "VmData"
+    let vmpte = "VmPTE"
+    let threads = "Threads"
+
+    (* there is also /proc/self/stat and /proc/self/statm, but we'd need to open and parse both *)
+    let status = define_fields ~path:"status"
+        [ count threads "Total number of threads used by %s"
+        ; count "FDSize" "Total number of file descriptors used by %s"
+        ; kib "VmSize" "Total amount of memory mapped by %s"
+        ; kib "VmLck" "Total amount of memory locked by %s"
+        ; kib "VmPin" "Total amount of memory pinned by %s"
+          (* VmRSS is inaccurate accoring to latest proc(5) *)
+        ; kib vmdata "Total amount of writable, non-shared and non-stack memory used by %s"
+        ; kib "VmStk" "Total amount of main stack memory used by %s"
+        ; kib vmpte "Total amount of page table entry memory used by %s"
+        ]
+
+    (* According to latest proc(5) these are slower, but provide more accurate information.
+       The RSS reported by other stat counters could be off depending on the number of threads. *)
+    let smaps_rollup = define_fields ~path:"smaps_rollup" [
+      kib "Rss" "Total amount of resident memory used by %s"
+    ]
+  end
+
+  let find key kv =
+    to_float (KV.get key kv)
+
+  let to_list kv =
+    KV.bindings kv |> List.rev_map snd
+end
+
+module GcStat = struct
+  open SimpleMetrics
+  open Gc
+
+  let ocaml_total =
+    let f = kib "ocaml_total" "Total OCaml memory used by %s" in
+    fun gc control ->
+    gc.heap_words + control.minor_heap_size
+    |> float |> words_to_kib |> f
+
+  let maybe_words name description v =
+    (* quick_stat would return a value of 0, which is not valid *)
+    v |> float |> words_to_kib |> kib ~min:0.001 name description
+
+  let compute (gc, control) =
```

Comment:
Not a good name - maybe `memory_churn` or `mem_allocation_rate_approx` and `mem_allocation_rate_prec` below?